utils/tqueue/tqueue.h

```diff
@@ -0,0 +1,90 @@
+#include <queue>
+#include <mutex>
+
+
+template<typename T>
+class tqueue {
+    std::queue<T> internal_queue;
+    std::mutex queue_write_mutex;
+
+    public: 
+
+    tqueue() : internal_queue(), queue_write_mutex() {}
+
+
+    T& front(){
+        return this->internal_queue.front();
+    }    
+
+
+    const T& front() const{
+        return this->internal_queue.front();
+    }
+
+
+    T& back(){
+        return this->internal_queue.back();
+    }
+
+
+    const T& back() const{
+        return this->internal_queue.back();
+    }
+
+
+    bool empty() const{
+        return this->internal_queue.empty();
+    }
+
+
+    size_t size() const{
+        return this->internal_queue.size();
+    }
+
+
+    void push( T&& value ){
+        this->queue_write_mutex.lock();
+        this->internal_queue.push(value);
+        this->queue_write_mutex.unlock();
+    }
+
+
+    void push_range( T&& rg ){
+        this->queue_write_mutex.lock();
+        this->internal_queue.push_range(rg);
+        this->queue_write_mutex.unlock();
+    }
+
+
+    decltype(auto) emplace( T&& t ){
```

Comment:
what is this decltype about?

Reply:
This is taken from the signature of the function of the standard library to wrap it around for TQueue. https://en.cppreference.com/w/cpp/container/queue/emplace
Do you suggest changing it or just using void here? I agree that decltype(auto) is not very pretty.

Reply:
ah right, didnt know emplace returns that since C++17

Reply:
keep it then